src/parsing_visualizer.cpp

```diff
@@ -0,0 +1,172 @@
+#include "include/parsing_visualizer.hpp"
+#include "include/parser.hpp"
+#include <vector>
+#include "gvpp.hpp"
+#include <iostream>
+#include <fstream>
+
+std::string stringify_op(const expression& expr, const char* sep, const char* delim, bool stmt, unsigned first, unsigned limit) {
+//    std::string result("\n");
+    std::string result(1, delim[0]);
+//    result += "\n";
+    const char *first_sep = "";
+    for (const auto &p: expr.get_parameters()) {
+        if (first) {
+            --first;
+            continue;
+        }
+        if (!limit--) break;
+        result += first_sep;
+        first_sep = sep;
+        result += stringify(p, stmt);
+    }
+    if (stmt) result += sep;
+    result += delim[1];
+    return result;
+}
+
+std::string expect_one_param(const expression &expr) {
+    if (expr.get_parameters().empty()) return "?";
+    else if (expr.get_parameters().size() == 1) {
+        return stringify(expr.get_parameters().front());
+    } else return stringify_op(expr, "??", "()");
+}
+
+std::string stringify(const expression& expr, bool stmt) {
```

Comment:
Swap the ternary operators with `if-else` statements.

Reply:
I prefer the ternaries over the ```if-else``` in this exact situation. I believe it looks clearer and is a good one-liner